pyboolnet/control_strategies.py

```diff
@@ -0,0 +1,486 @@
+    Checks whether the subspace *candidate* is a control strategy for *target* using model checking.
+
+    **arguments**:
+        * *primes*: prime implicants.
+        * *candidate*: subspace.
+        * *target*: list of subspaces defining the target subset.
+        * *update*: type of update, either *"synchronous"*, *"asynchronous"* or *"mixed"*.
+        * *silent*: if True, it does not print infos to screen. Default value: False.
+
+    **returns**:
+        * True if the *candidate* is a control strategy by completeness for *target*, False otherwise.
+    """
+
+    if type(target) == dict:
+        print("The target must be a list of subspaces.")
+        return
+
+    perc = find_constants(primes=percolate(primes=primes, add_constants=candidate, copy=True))
+    new_primes = fix_components_and_reduce(primes, perc, keep_vars=list(set([item for subs in target for item in subs])))
+    tsmin = compute_trap_spaces(new_primes, "min", max_output=max_output_trapspaces)
+
+    if not control_is_valid_in_trap_spaces(new_primes, tsmin, target, update):
+        return False
+
+    if run_control_query(new_primes, target, update):
+        if not silent:
+            print("Intervention (by CTL formula):", candidate)
+            print("Percolation:", perc)
+        return True
+
+    return False
+
+
+def find_necessary_interventions(primes: dict, target: List[dict]) -> dict:
+    """
+    Find name and values of the inputs and constants from *primes* that are fixed in the *target*.
+
+    **arguments**:
+        * *primes*: prime implicants
+        * *target*: list of subspaces
+
+    **returns**:
+        * *selected_vars*: Names and values that are inputs or constants in *primes* and are fixed in the *target*.
+    """
+
+    selected_vars = dict()
+    candidates = find_inputs(primes) + list(find_constants(primes).keys())
+    for x in candidates:
+        if all(x in y.keys() for y in target):
+            if all(y[x] == z[x] for y in target for z in target):
+                selected_vars[x] = target[0][x]
+    return selected_vars
+
+
+def find_common_variables_in_cs(primes: dict, target: List[dict]) -> dict:
+    """
+    Find name and values of the constants from *primes* that are fixed to a different value in the *target*.
+
+    **arguments**:
+        * *primes*: prime implicants
+        * *target*: list of subspaces
+
+    **returns**:
+        * Names and values that are constants in *primes* and are fixed to a different value in the *target*.
+    """
+
+    inputs_consts_common_in_subs = find_necessary_interventions(primes, target)
+    right_consts = [x for x in find_constants(primes) if x in inputs_consts_common_in_subs.keys() and inputs_consts_common_in_subs[k] == consts[x]]
+    return dict((k, inputs_consts_common_in_subs[k]) for k in inputs_consts_common_in_subs.keys() if k not in right_consts)
+
+
+def is_control_strategy(primes: dict, candidate: dict, target: List[dict], update: str, max_output: int = 1000000):
```

Comment:
type hint of returned value

(the reason is the modern IDEs can detect if the user has a mistake in using the return value (if you specify it)